### ISSUE TITLE
Show videos as well [$75]

### ISSUE THREAD

**Issue by oparoz** [nextcloud/gallery#47]
_From @oparoz on June 27, 2015 13:51_

Showing the thumbnails is easy. A designer just needs to design a play button overlay to indicate this is a video.
Integrating the 2 players (image and video) in the slideshow so that they look like one and the same is a bit harder.
The real hurdle here is that most video formats don't play in a browser without using Flash.
The best solution would be to transcode videos, kind of like what we do with images, we create a thumbnail which can be viewed by the browser.

I definitely think this should be optional.

If you want to read more comments:
https://github.com/owncloud/apps/issues/1437

### Caveats / Risks

* It's not possible to play videos located on remote shares. We need to be able to scan the file's header to be able to determine if we can play it

## Phase 1

* Analyse files with video media types in order to be able to only pick the ones we can support via an HTML5 video player. This requires some external dependencies, per example: https://github.com/JamesHeinrich/getID3
* Make video support optional because of the required dependencies

## Phase 2

* Add supported files to the list of files located in the scanned album
* Add a play button overlay
* Pick an HTML5 player
* Add the player to the slideshow

## Phase 3

* Style the player
* Allow seeking (only works if core supports it)

## Phase 4 (out of scope)

* Transcode formats

_Copied from original issue: owncloud/gallery#194_

**Comment by oparoz** [comment]
_From @jancborchardt on June 28, 2015 0:55_

For the play button overlap, you can simply use a div with .icon-view-play with some darker half-transparent background. :) I will do detail styling.

**Comment by oparoz** [comment]
_From @jancborchardt on November 5, 2015 16:11_

cc @karlitschek @DeepDiver1975 FYI

**Comment by oparoz** [comment]
_From @karlitschek on November 5, 2015 19:25_

nice

**Comment by oparoz** [comment]
_From @jancborchardt on November 9, 2015 17:33_

cc @VicDeo of our current [videoviewer](https://github.com/owncloud/apps/tree/master/files_videoviewer) app on this, as well as @Instagit of this awesome new video player app: https://github.com/Instagit/videojs-sublime

**Comment by oparoz** [comment]
The problems rarely come from the players, but from the fact that we usually have no idea which files will play or leave users hanging...
So in terms of usability, phase 1 is crucial, but requires external dependencies like mplayer.

**Comment by oparoz** [comment]
_From @jancborchardt on November 11, 2015 16:21_

@oparoz well the player of @Instagit has a simple message that the file type is not supported. I’d say that’s fine for the first pass. Because the video is shown and you don’t wonder where it is, and you are informed why it can not be played.

**Comment by oparoz** [comment]
I don't think it's fine in the Files app as the user will have a hit and miss experience. Videos with thumbnails might not play. That's very confusing. Videos without thumbnails (probably rare) might never be clicked on.

It's worse in Gallery as we would be showing thumbnails of unplayable media, leading to frustration.

I've been there. Installed video players, uploaded a mixed collection and couldn't play most of it. The players only support very specific formats.

**Comment by oparoz** [comment]
_From @jancborchardt on November 11, 2015 17:9_

Ok – but the situation would not be any different compared to now. So changing the player to a better looking and working one (on mobile as well) would be a great improvement already.

The mplayer stuff and extra dependencies is a further enhancement, but a slightly different tangent.

**Comment by oparoz** [comment]
Yes, but that has nothing to do with Gallery. People install an app and get a hit and miss experience. After that, they may decide to remove the app.

If Gallery claims to support videos and only offers patchy support, we're going to get lots of requests from people.

**Comment by oparoz** [comment]
_From @devurandom on November 11, 2015 20:56_

If said player (@jancborchardt's recommendation) can display a message, won't ownCloud/Gallery also be able to determine whether the file can be played or not? And then display the thumbnail/preview greyed out, with some explanatory text in front of it?

It has to show all possible thumbnails anyway, because it's IMO supposed to show me all files of a certain type (i.e. videos and pictures), not just the ones it can play (otherwise I'd wonder why Gallery completely hides a certain file - without ever knowing the reason). So the server generates the list as usual, and Gallery just desaturises the preview/thumbnail using some JS script on the client, and maybe adds a text somewhere (for large previews that can well be the preview image itself).

**Comment by oparoz** [comment]
_From @Instagit on November 12, 2015 8:26_

The best you can do without starting the video is try to guess if the file can be played with *canPlayType()*: http://www.leanbackplayer.com/test/h5mt.html

With only the file extension, there's no way to say for sure if the file can be played, because that depends on the codecs used.

With external tools that detect the codecs like mplayer mentioned here or FFmpeg, the guess could come close to 100% precision, but I think that would be a lot of work to implement and could also slow down loading times when there are many videos in the gallery.

**Comment by oparoz** [comment]
> With only the file extension, there's no way to say for sure if the file can be played, because that depends on the codecs used.

Exactly and that's the source of a lot of frustration and misunderstanding because people will say: "I have this mkv or MP4 and the player gives me nothing/an empty video/no audio/etc."

> but I think that would be a lot of work to implement

I don't think so, but it does require to use unsafe PHP methods.

> and could also slow down loading times when there are many videos in the gallery.

Not if we store the format the first time we encounter a new video. It's similar to the current effort to support EXIF meta data extracted from images.

> If said player can display a message, won't ownCloud/Gallery also be able to determine whether the file can be played or not?

Yes, but only after clicking on the video and as mentioned earlier, you may get a thumbnail, but the player won't show you the video. Then it's possible to grey-out the video, but the app will forget about it and you'll click again the next time you launch the app.

> It has to show all possible thumbnails anyway, because it's IMO supposed to show me all files of a certain type (i.e. videos and pictures), not just the ones it can play (otherwise I'd wonder why Gallery completely hides a certain file - without ever knowing the reason)

I agree. Gallery displays broken JPEGs so that the user can download them. We could do the same with videos, showing the video icon instead of a thumbnail and letting the user try his luck.
If we ever implement a way to detect the codec, then we can enable thumbnail previews.

**Comment by oparoz** [comment]
@Instagit - Does you player detect the VLC plugin and use it?

**Comment by oparoz** [comment]
Found it:
https://github.com/Afterster/videojs-vlc

**Comment by oparoz** [comment]
_From @devurandom on November 13, 2015 15:55_

> > If said player can display a message, won't ownCloud/Gallery also be able to determine whether the file can be played or not?
> 
> Yes, but only after clicking on the video and as mentioned earlier, you may get a thumbnail, but the player won't show you the video. Then it's possible to grey-out the video, but the app will forget about it and you'll click again the next time you launch the app.

You could store the result of that examination, as long as the player is able to give you a list of codecs used in the file, or similar.

I assume one should not generally store "playable: yes/no" in the database anyway, as each browser might support a different list of codecs.
